api/routers/v1/timelog.py

```diff
@@ -66,10 +67,52 @@ async def add_template(
     else:
         if template.user_id and template.user_id != current_user.id:
             raise HTTPException(status_code=403, detail="You are not authorized to create templates for this user")
+    if template.task_type:
+        task_type_valid = TaskTypeService(db).slug_is_valid(template.task_type)
+        if not task_type_valid:
+            raise HTTPException(status_code=422, detail=f"Task type {template.task_type} does not exist.")
     result = TemplateService(db).create_template(template)
     return result
 
 
+@router.put("/templates", response_model=TemplateSchema)
+async def update_template(
+    template_id: int,
+    template: TemplateUpdateSchema,
+    current_user=Depends(get_current_user),
+    db: Session = Depends(get_db),
+):
+    """
+    Update a template with any of the following:
+
+    - **name**: each template must have a name
+    - **story**: the task story
+    - **description**: the task description
+    - **task type**: the task type
+    - **user id**: the user id (global templates should leave this null; user template should fill)
+    - **project id**: the project id
+    - **is global**: whether or not this template is global for all users (required)
+    - **start time**: the task start time
+    - **end time**: the task end time
+    \f
+    :param item: User input.
+    """
+    existing_template = TemplateService(db).get_template(template_id)
+    if not existing_template:
+        raise HTTPException(status_code=404, detail=f"Template with id {template_id} not found")
+    if template.is_global and "manager" not in current_user.roles:
+        raise HTTPException(status_code=403, detail="You are not authorized to update global templates")
+    else:
+        if template.user_id and template.user_id != current_user.id:
+            raise HTTPException(status_code=403, detail="You are not authorized to update templates for this user")
+    if template.task_type:
+        task_type_valid = TaskTypeService(db).slug_is_valid(template.task_type)
+        if not task_type_valid:
+            raise HTTPException(status_code=422, detail=f"Task type {template.task_type} does not exist.")
```

Comment:
This is the same logic from creating a template, should we move it to a method?

Reply:
I moved it to its own method and did a small refactor: I had a TaskValidate object, but moved it to validation.py and repurposed it as a general validator object that we could use for anything.